cpp/src/tests/sort/sort_test.cu

```diff
@@ -0,0 +1,443 @@
+{
+  cudaStream_t stream{nullptr};
+  
+  uint64_t           *d_input[MAX_NUM_GPUS];
+  uint64_t           *d_input_values[MAX_NUM_GPUS];
+  uint64_t           *d_output[MAX_NUM_GPUS];
+  uint64_t           *d_output_values[MAX_NUM_GPUS];
+  unsigned long long  h_input_offsets[3];
+  unsigned long long  h_output_offsets[3];
+
+  const int num_elements = 10000000;
+  const int seed = 43;
+  int n_gpus = 0;
+
+  CUDA_RT_CALL(cudaGetDeviceCount(&n_gpus));
+
+  for (int i = 0 ; i < (n_gpus + 1) ; ++i)
+    h_input_offsets[i] = i * num_elements;
+
+  omp_set_num_threads(n_gpus);
+
+  generate_random(d_input, d_input_values, h_input_offsets, n_gpus, seed, stream);
+
+  //
+  //  Initialize Peer-to-peer communications
+  //
+  cusort::initialize_snmg_communication(n_gpus);
+
+  // NOTE:  could vary numBins, binScale, useThrust
+  cusort::sort_key_value(d_input, d_input_values, h_input_offsets,
+                         d_output, d_output_values, h_output_offsets,
+                         n_gpus);
+
+  verify_sorted_order(d_output, d_output_values, h_output_offsets, n_gpus, stream, true);
+
+  for (int i = 0 ; i < n_gpus ; ++i) {
+    cudaSetDevice(i);
+
+    EXPECT_EQ(RMM_FREE(d_input[i], stream), RMM_SUCCESS);
+    EXPECT_EQ(RMM_FREE(d_output[i], stream), RMM_SUCCESS);
+  }
+}
+
+TEST_F(SortTest, Random10MPerDevice_uint32_t)
+{
+  cudaStream_t stream{nullptr};
+  
+  uint32_t           *d_input[MAX_NUM_GPUS];
+  uint32_t           *d_input_values[MAX_NUM_GPUS];
+  uint32_t           *d_output[MAX_NUM_GPUS];
+  uint32_t           *d_output_values[MAX_NUM_GPUS];
+  unsigned long long  h_input_offsets[3];
+  unsigned long long  h_output_offsets[3];
+
+  const int num_elements = 10000000;
+  const int seed = 43;
+  int n_gpus = 0;
+
+  CUDA_RT_CALL(cudaGetDeviceCount(&n_gpus));
+
+  for (int i = 0 ; i < (n_gpus + 1) ; ++i)
+    h_input_offsets[i] = i * num_elements;
+
+  omp_set_num_threads(n_gpus);
+
+  generate_random(d_input, d_input_values, h_input_offsets, n_gpus, seed, stream);
+
+  //
+  //  Initialize Peer-to-peer communications
+  //
+  cusort::initialize_snmg_communication(n_gpus);
+
+  // NOTE:  could vary numBins, binScale, useThrust
+  cusort::sort_key_value(d_input, d_input_values, h_input_offsets,
+                         d_output, d_output_values, h_output_offsets,
+                         n_gpus);
+
+  verify_sorted_order(d_output, d_output_values, h_output_offsets, n_gpus, stream, true);
+
+  for (int i = 0 ; i < n_gpus ; ++i) {
+    cudaSetDevice(i);
+
+    EXPECT_EQ(RMM_FREE(d_input[i], stream), RMM_SUCCESS);
+    EXPECT_EQ(RMM_FREE(d_output[i], stream), RMM_SUCCESS);
+  }
+}
+
+TEST_F(SortTest, Random100MPerDevice_uint64_t)
+{
+  cudaStream_t stream{nullptr};
+  
+  uint64_t           *d_input[MAX_NUM_GPUS];
+  uint64_t           *d_input_values[MAX_NUM_GPUS];
+  uint64_t           *d_output[MAX_NUM_GPUS];
+  uint64_t           *d_output_values[MAX_NUM_GPUS];
+  unsigned long long  h_input_offsets[3];
```

Comment:
unsigned long long  h_input_offsets[MAX_NUM_GPUS+1];
  unsigned long long  h_output_offsets[MAX_NUM_GPUS+1];

Reply:
why use "unsigned long long" versus "uint64_t" ?

Reply:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicaddA lot of the class is templated and uses atomicAdds in various locations. unit64_t can cause compile time warnings and errors.
--

Reply:
oops, I do remember that about atomics.  good catch

Reply:
I tried to minimize the places where unsigned long long was used to only those places that get used by atomics.  It would be nice to have the cuda atomic add fixed to support all of the various spellings of the type names properly.

I will shortly push a fix to the hardcoded 3... that was careless.